### ISSUE TITLE
cypher toString(null) + regex throws exception in neo4j 4.0.8 and 4.1.2

### ISSUE THREAD

**Issue by mmopitz** [neo4j/neo4j#12592]
-  I tested with Neo4j versions: 4.0.0 & 4.0.8 and  4.1.0 & 4.1.2 
- Operating system: ubuntu 18.04
- Neo4j Desktop 1.3.8

I'm currently migrating from neo4j 3.5 to neo4j 4.0 .

I found that the behavior of toString(null) + a regular expression throws an exception in neo4j 4.0 and 4.1 . In neo4j 3.5 it worked just fine. I don't know if that's intended.

### Steps to reproduce 

Note: I used neo4j desktop to create the databases and did the testing with the neo4j browser
1. create a neo4j database version 4.1.2
2. open up neo4j browser
3. create testdata (one person with a name. And one person without: 
`create (n:Person {name : 'a'})
create (n2:Person)`
4. run following query:
`match (n:Person)
return n.name ,tostring(n.name) =~ '(?i).*a.*'`

### Outcome:
I get following error: 
```
Neo.ClientError.Statement.TypeError
Expected a String, Number, Boolean, Temporal or Duration, got: NO_VALUE
```

### Expected Outcome:
```
╒════════╤═══════════════════════╕
│"n.name"│"n.name =~ '(?i).*a.*'"│
╞════════╪═══════════════════════╡
│"a"     │true                   │
├────────┼───────────────────────┤
│null    │null                   │
└────────┴───────────────────────┘
```
Note: In neo4j 3.5 the query evaluates to the expected outcome

### Stacktrace
I found following stacktrace in the query.log
```
2020-09-28 09:45:36.211+0000 ERROR id:1075 - 1 ms: bolt-session	bolt	neo4j-javascript/4.1.1		client/127.0.0.1:33596	server/127.0.0.1:7687>	neo4j - neo4j - match (n:Person) return n.name ,toString(n.name) =~ '(?i).*a.*' - {} - {type: 'user-direct', app: 'neo4j-browser_v4.1.2'} Expected a String, Number, Boolean, Temporal or Duration, got: NO_VALUE
org.neo4j.exceptions.ParameterWrongTypeException: Expected a String, Number, Boolean, Temporal or Duration, got: NO_VALUE
	at org.neo4j.cypher.operations.CypherFunctions.toString(CypherFunctions.java:1161)
	at org.neo4j.codegen.OperatorTask23.compiledOperate(Unknown Source)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.CompiledTask.operateWithProfile(OperatorCodeGenBaseTemplates.scala:179)
	at org.neo4j.cypher.internal.runtime.pipelined.PipelineTask.executeOperators(PipelineTask.scala:58)
	at org.neo4j.cypher.internal.runtime.pipelined.PipelineTask.executeWorkUnit(PipelineTask.scala:48)
	at org.neo4j.cypher.internal.runtime.pipelined.Worker.executeTask(Worker.scala:128)
	at org.neo4j.cypher.internal.runtime.pipelined.Worker.workOnQuery(Worker.scala:84)
	at org.neo4j.cypher.internal.runtime.pipelined.execution.CallingThreadExecutingQuery.request(CallingThreadExecutingQuery.scala:30)
	at org.neo4j.cypher.internal.PipelinedRuntime$PipelinedRuntimeResult.request(PipelinedRuntime.scala:347)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.request(StandardInternalExecutionResult.scala:88)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.request(ClosingExecutionResult.scala:135)
	at org.neo4j.bolt.runtime.AbstractCypherAdapterStream.handleRecords(AbstractCypherAdapterStream.java:105)
	at org.neo4j.bolt.v3.messaging.ResultHandler.onPullRecords(ResultHandler.java:41)
	at org.neo4j.bolt.v4.messaging.PullResultConsumer.consume(PullResultConsumer.java:42)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State.consumeResult(TransactionStateMachine.java:511)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State$2.streamResult(TransactionStateMachine.java:355)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine.streamResult(TransactionStateMachine.java:92)
	at org.neo4j.bolt.v4.runtime.InTransactionState.processStreamResultMessage(InTransactionState.java:73)
	at org.neo4j.bolt.v4.runtime.AbstractStreamingState.processUnsafe(AbstractStreamingState.java:49)
	at org.neo4j.bolt.v4.runtime.InTransactionState.processUnsafe(InTransactionState.java:60)
	at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48)
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:143)
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:91)
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:90)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:151)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:240)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:175)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:165)
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:212)
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$2(ExecutorBoltScheduler.java:195)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### Some further digging
Looks like the problem doesn't occur if I don't apply a regex filter:
E.g following query works fine : 
`match (n:Person) return n.name ,toString(n.name)='a'` 
works just fine

**Comment by mnd999** [comment]
Thanks for the report, we'll look into it and let you know if we have a fix.

**Comment by loveleif** [comment]
Thanks again for reporting this @mmopitz! The fix has been merged to 4.2, 4.1 and 4.0 branches. It will be part of upcoming releases of these versions.

**Comment by mmopitz** [comment]
Thanks a lot :-)
